src/server/pps/server/defaults.go

```diff
@@ -307,21 +307,25 @@ func init() {
 // JSON-encoded ClusterDefaults) and user spec (a JSON-encoded
 // CreatePipelineRequest) by merging the user spec into the cluster defaults.
 // It returns the effective spec as both JSON and a CreatePipelineRequest.
-func makeEffectiveSpec(clusterDefaultsJSON, userSpecJSON string) (string, *pps.CreatePipelineRequest, error) {
+func makeEffectiveSpec(clusterDefaultsJSON, projectDefaultsJSON, userSpecJSON string) (string, *pps.CreatePipelineRequest, error) {
 	type wrapper struct {
 		CreatePipelineRequest json.RawMessage `json:"createPipelineRequest"`
 	}
 	effectiveClusterDefaults, err := jsonMergePatch(builtInDefaultsJSON, clusterDefaultsJSON, clusterDefaultsCanonicalizer)
 	if err != nil {
 		return "", nil, errors.Wrapf(err, "could not merge cluster defaults %s into built-in defaults %s", clusterDefaultsJSON, builtInDefaultsJSON)
 	}
+	effectiveProjectDefaults, err := jsonMergePatch(effectiveClusterDefaults, projectDefaultsJSON, clusterDefaultsCanonicalizer)
```

Comment:
if jsonMergePatch() took a variadic for patches, this whole function could be expressed in one call with arguments `(builtInDefaultsJSON, clusterDefaultsJSON, projectDefaultsJSON,  string(userWrapper), clusterDefaultsCanonicalizer)`

Reply:
I considered that, but my concern is that a call site might remain, or be introduced, which inadvertently omits one or more JSON strings.

Were this an exported API meant for general use, that would be more appealing, but since it’s internal I think that it makes sense for it to be less general and more error-proof.